### ISSUE TITLE
inconsistant result with differect thread_count setted

### ISSUE THREAD

**Issue by xq-yu** [catboost/catboost#2804]
When I trained a CatBoost classifier(PS: regressor model is OK) model in two different environments using the default thread_count parameter and exactly the same code, I found that the results were different. I realized this discrepancy was due to the different numbers of CPU cores in the two environments. After explicitly setting the thread_count parameter, it was proven that thread_count can indeed affect the results. 

Here is the model configuration:

params = {'loss_function':'Logloss',
          'eval_metric':'AUC',
          'cat_features': data['feature_cat'],
          'early_stopping_rounds': 100,
          'verbose': 100,
          'random_seed': 43,
          'depth':6, 
          'learning_rate':0.02,
          'n_estimators':20000,
          'subsample':0.66,
          'colsample_bylevel':0.8,
          'l2_leaf_reg':1,
          'thread_count':2  #result would change with this param
         }
catbst_cls = CatBoostClassifier(**params)

**Comment by andrey-khropov** [comment]
How different were the results? Was there a significant difference in evalution metrics between training runs with different `thread_count` values.

Also, please specify your environment in more detail:

CatBoost version,
CPU(s) model,
Operating system

**Comment by manlizhao** [comment]
@andrey-khropov Hi, actually I have a question about the parameter thread count, how should I adjust the parameter to optimize the training?

**Comment by manlizhao** [comment]
@andrey-khropov  thread_count = -1 doesn't seem to allow

**Comment by andrey-khropov** [comment]
> [@andrey-khropov](https://github.com/andrey-khropov) Hi, actually I have a question about the parameter thread count, how should I adjust the parameter to optimize the training?

If you just train a single model at the time on a single host you usually don't need to specify this parameter explicitly - CatBoost will just use all the available CPU cores by default.

It is useful only if you want to restrict the number of used CPU cores for a single model training if, for example, you want to train several models at the same time in parallel.

**Comment by andrey-khropov** [comment]
> [@andrey-khropov](https://github.com/andrey-khropov) thread_count = -1 doesn't seem to allow

I don't understand what you mean by that. Can you elaborate?

**Comment by manlizhao** [comment]
> > [@andrey-khropov](https://github.com/andrey-khropov) thread_count = -1 doesn't seem to allow
> 
> I don't understand what you mean by that. Can you elaborate?

I saw this comment, so I tried to change it to -1. But it gave me error.

https://github.com/catboost/catboost/issues/795#issuecomment-588032965

**Comment by xq-yu** [comment]
> How different were the results? Was there a significant difference in evalution metrics between training runs with different `thread_count` values?
> 
> Also, please specify your environment in more detail:
> 
> CatBoost version, CPU(s) model, Operating system

firstly, i found this problem in 2 environment, then i checked in the same environment with different thread_count(2,4), the problem accured again and the auc metric with 5-fold cross validation is 0.7612 and 0.7609. 


---------------------------------------
catboost==1.2.7

Distributor ID: Ubuntu
Description:    Ubuntu 20.04.6 LTS
Release:        20.04
Codename:       focal

CPU:12th Gen Intel(R) Core(TM) i5-12400F
---------------------------------------

**Comment by andrey-khropov** [comment]
> > > [@andrey-khropov](https://github.com/andrey-khropov) thread_count = -1 doesn't seem to allow
> > 
> > 
> > I don't understand what you mean by that. Can you elaborate?
> 
> I saw this comment, so I tried to change it to -1. But it gave me error.

What kind of error exactly?
